### PR TITLE
Transform travis config for npm usage

### LINKED ISSUE #186: Update .travis.yml files to use --npm and yarn
See https://github.com/PolymerElements/marked-element/pull/77

### PR DESCRIPTION
**Pull request by azakus**
Fixes #186

### CODE REVIEW

**Comment by stramel** [comment]
Hmmm @azakus your changes just overwrite travis.yml file's sections. #276 doesn't

File: src/travis-config.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  const travisConfig = safeLoad(travisBlob) as Partial<TravisConfig>;
+
+  // use 'yarn' to install
+  travisConfig.install = 'yarn';
+
+  // cache yarn directories
+  travisConfig.cache = 'yarn';
+
+  // install polymer-cli
+  travisConfig.before_script = 'npm install -g polymer-cli';
```
Comment (FredKSchott) [inline src/travis-config.ts]: If we're using yarn, best to do do`yarn global add polymer-cli` here

File: src/travis-config.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as fse from 'fs-extra';
+import {safeDump, safeLoad} from 'js-yaml';
+import * as path from 'path';
+
+const TRAVIS_CONFIG = '.travis.yml';
```
Comment (FredKSchott) [inline src/travis-config.ts]: nit: we generally don't follow the CAPS_LOCK_SNAKE_CASE for constants, `travisConfig` is fine

**Comment by azakus** [comment]
@stramel Ah yeah, very true.

File: src/travis-config.ts (hunk 1 above)
Comment (justinfagnani) [reply src/travis-config.ts]: why? npm for global things and yarn for local things is fine, and works better in my experience.

File: src/travis-config.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * @param outDir Root path of output travis config
+ */
+export async function transformTravisConfig(
+    inDir: string, outDir: string): Promise<void> {
+  const inTravisPath = path.join(inDir, TRAVIS_CONFIG);
+  const travisBlob = (await fse.readFile(inTravisPath)).toString();
+  const travisConfig = safeLoad(travisBlob) as Partial<TravisConfig>;
+
+  // use 'yarn' to install
+  travisConfig.install = 'yarn';
```
Comment (justinfagnani) [inline src/travis-config.ts]: I wouldn't use yarn unless we're using --flat.

**Comment by justinfagnani** [comment]
So... @azakus and @stramel 

We now have 2 PRs open on this issue. I'm assuming that Dan just didn't see the other PR. I like some things about each. I like that this is in a separate file with the typings for Travis configs in it, and that Michael's attempts to patch up commands. Who wants to update their PR with aspect of the others?

**Comment by azakus** [comment]
I'm happy to take the good bits of @stramel's PR into this one. Is that ok with you @stramel? I'll make sure to note your contributions :)

**Comment by stramel** [comment]
That would be great @azakus! Thank you!

File: fixtures/packages/iron-icon/expected/.travis.yml [hunk 4]
```diff
@@ -22,8 +21,8 @@ addons:
     packages:
       - google-chrome-stable
 script:
-  - xvfb-run polymer test
+  - xvfb-run polymer-test --npm
```
Comment (justinfagnani) [inline fixtures/packages/iron-icon/expected/.travis.yml]: why is there a dash in `polymer-test` now?

File: fixtures/packages/iron-icon/expected/.travis.yml [hunk 5]
```diff
@@ -22,8 +21,8 @@ addons:
...
   - >-
-    if [ "${TRAVIS_PULL_REQUEST}" = "false" ]; then polymer test -s 'default';
-    fi
+    if [ "${TRAVIS_PULL_REQUEST}" = "false" ]; then polymer-test --npm -s
+    'default'; fi
```
Comment (justinfagnani) [inline fixtures/packages/iron-icon/expected/.travis.yml]: is there a way to get js-yaml to not reformat the file so diffs are more targeted?

File: src/travis-config.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  install?: string|string[];
+  script?: string|string[];
+  cache?: string|Array<string|{directories: string[]}|{[key: string]: boolean}>;
+}
+
+// modify travis scripts, credit to @stramel for the modifications to make
+function addNPMFlag(scripts: string[]): string[] {
+  return scripts.map((script) => {
+    if (script.indexOf('polymer test') > -1 && script.indexOf('--npm') === -1) {
+      script = script.replace('polymer test', 'polymer-test --npm');
```
Comment (justinfagnani) [inline src/travis-config.ts]: `'polymer-test --npm'` -> `'polymer test --npm'`

File: src/travis-config.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    return [];
+  } else if (typeof yamlPart === 'string') {
+    return [yamlPart];
+  } else {
+    return yamlPart;
+  }
+}
+
+/**
+ * Modify a project's travis config to use `yarn`
```
Comment (justinfagnani) [inline src/travis-config.ts]: Update comment? We shouldn't be changing to yarn

File: src/travis-config.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ */
+export async function transformTravisConfig(
+    inDir: string, outDir: string): Promise<void> {
+  const inTravisPath = path.join(inDir, travisConfigFile);
+
+  if (!await fse.pathExists(inTravisPath)) {
+    return;
+  }
+
+  const travisBlob = (await fse.readFile(inTravisPath)).toString();
```
Comment (justinfagnani) [inline src/travis-config.ts]: If you pass an encoding, `readFile` will resolve a string.

File: fixtures/packages/iron-icon/expected/.travis.yml (hunk 4 above)
Comment (azakus) [reply fixtures/packages/iron-icon/expected/.travis.yml]: Oops, my replace is wrong

File: src/travis-config.ts (hunk 6 above)
Comment (azakus) [reply src/travis-config.ts]: Same as first comment.

File: fixtures/packages/iron-icon/expected/.travis.yml (hunk 5 above)
Comment (azakus) [reply fixtures/packages/iron-icon/expected/.travis.yml]: Maybe, I'll take a look

File: src/travis-config.ts (hunk 7 above)
Comment (azakus) [reply src/travis-config.ts]: Yeah that's leftover

File: fixtures/packages/iron-icon/expected/.travis.yml (hunk 5 above)
Comment (azakus) [reply fixtures/packages/iron-icon/expected/.travis.yml]: Ok, I can turn off lineWidth checking, and make sure that single item arrays are turned into a string.

Comment (justinfagnani) [reply fixtures/packages/iron-icon/expected/.travis.yml]: awesome!

**Comment by justinfagnani** [comment]
@azakus I forgot one thing, we need to also add the --module-resolution=node flag for `polymer test`

**Comment by azakus** [comment]
Have a PR for that in #360

**Pull request merged**